near-sdk-macros/src/core_impl/utils/mod.rs

```diff
@@ -73,3 +75,62 @@ pub(crate) fn extract_vec_type(ty: &Type) -> Option<&Type> {
         _ => None,
     }
 }
+
+fn _sanitize_self(typ: TokenStream2, replace_with: &TokenStream2) -> TokenStream2 {
+    let trees = typ.into_iter().map(|t| match t {
+        TokenTree::Ident(ident) if ident == "Self" => {
+            let replace_with = replace_with
+                .clone()
+                .into_iter()
+                .map(|mut t| {
+                    t.set_span(ident.span());
+                    t
+                })
+                .collect();
+            TokenTree::Group(Group::new(Delimiter::None, replace_with))
+        }
+        TokenTree::Group(group) => {
+            let stream = _sanitize_self(group.stream(), replace_with);
+            TokenTree::Group(Group::new(group.delimiter(), stream))
+        }
+        rest => rest,
+    });
+    trees.collect()
+}
+
+pub fn sanitize_self(typ: &Type, replace_with: &TokenStream2) -> syn::Result<Type> {
+    syn::parse2(_sanitize_self(quote! { #typ }, replace_with)).map_err(|original| {
```

Comment:
Sanitizing `TokenStream` instead of `Type` is a nice trick. I remember trying to sanitize `Self` in the past, and it required so much boilerplate to traverse all `Type` variants.

Reply:
Yeah, nightmarish 😨